File: packages/core/src/core/skill-loader.ts
/**
 * Skill Loader — loads and validates skills from multiple sources
 *
 * Implements agentskills.io specification with proper YAML parsing and Zod validation.
 *
 * @see https://agentskills.io/specification
 */

import fs from 'fs';
import path from 'path';
import YAML from 'js-yaml';
import { z } from 'zod';
import { SkillDefinition, SkillFrontmatter, ParsedSkill, BundledResources } from './types';
import { parseSkillSections } from './skill-content-parser';
import { getGlobalMatimoLogger } from '../logging';
import { MatimoError, ErrorCode } from '../errors/matimo-error';

// ─── Name Validation ─────────────────────────────────────────────────────────

const VALID_NAME_PATTERN = /^[a-z0-9]([a-z0-9-]*[a-z0-9])?$/;
const CONSECUTIVE_HYPHENS = /--/;
const MAX_NAME_LENGTH = 64;

function validateSkillName(name: string): { valid: boolean; error?: string } {
  if (!name || name.trim().length === 0) {
    return { valid: false, error: 'Skill name is required' };
  }

  if (name.length > MAX_NAME_LENGTH) {
    return {
      valid: false,
      error: `Skill name must be at most ${MAX_NAME_LENGTH} characters`,
    };
  }

  if (!VALID_NAME_PATTERN.test(name)) {
    return {
      valid: false,
      error:
        'Skill name must contain only lowercase letters, numbers, and hyphens, and must not start or end with a hyphen',
    };
  }

  if (CONSECUTIVE_HYPHENS.test(name)) {
    return { valid: false, error: 'Skill name must not contain consecutive hyphens' };
  }

  return { valid: true };
}

// ─── YAML Parser & Validation ──────────────────────────────────────────────

/**
 * Parser schema for skill frontmatter (using Zod for runtime validation)
 */
const FrontmatterSchema = z.object({
  name: z.string().min(1, 'name is required'),
  description: z.string().min(1, 'description is required').max(1024),
  version: z.string().optional(),
  license: z.string().optional(),
  compatibility: z.string().max(500).optional(),
  'allowed-tools': z.union([z.string(), z.array(z.string())]).optional(),
  metadata: z.record(z.string(), z.string()).optional(),
});

/**
 * Parse YAML frontmatter from SKILL.md content
 */
export function parseSkillContent(content: string): ParsedSkill & { error?: string } {
  if (!content || !content.startsWith('---')) {
    return {
      error: 'Skill content must start with YAML frontmatter (---)',
      frontmatter: { name: '', description: '' },
      body: '',
      raw: content,
    };
  }

  const endIndex = content.indexOf('---', 3);
  if (endIndex === -1) {
    return {
      error: 'Skill content must have closing YAML frontmatter (---)',
      frontmatter: { name: '', description: '' },
      body: '',
      raw: content,
    };
  }

  const frontmatterBlock = content.substring(3, endIndex).trim();
  const body = content.substring(endIndex + 3).trim();

  let parsed: Record<string, unknown>;
  try {
    parsed = (YAML.load(frontmatterBlock) as Record<string, unknown>) || {};
  } catch (e) {
    return {
      error: `Failed to parse YAML frontmatter: ${(e as Error).message}`,
      frontmatter: { name: '', description: '' },
      body,
      raw: content,
    };
  }

  // Normalize allowed-tools: convert space-delimited string to array
  if (parsed['allowed-tools'] && typeof parsed['allowed-tools'] === 'string') {
    parsed['allowed-tools'] = parsed['allowed-tools'].split(/\s+/);
  }

  // Validate with Zod
  const validationResult = FrontmatterSchema.safeParse(parsed);
  if (!validationResult.success) {
    const errors = validationResult.error.issues
      .map((e: z.ZodIssue) => `${e.path.join('.')}: ${e.message}`)
      .join('; ');
    return {
      error: `Frontmatter validation failed: ${errors}`,
      frontmatter: { name: '', description: '' },
      body,
      raw: content,
    };
  }

  const frontmatter: SkillFrontmatter = {
    name: validationResult.data.name,
    description: validationResult.data.description,
    version: validationResult.data.version,
    license: validationResult.data.license,
    compatibility: validationResult.data.compatibility,
    'allowed-tools': validationResult.data['allowed-tools'] as string[] | undefined,
    metadata: validationResult.data.metadata,
  };

  // Parse body into structured sections for selective context loading
  const parsedContent = parseSkillSections(body);

  return {
    frontmatter,
    body,
    raw: content,
    sections: parsedContent.sections,
    totalTokens: parsedContent.totalTokens,
  };
}

// ─── Bundled Resources Discovery ───────────────────────────────────────────

function listBundledResources(skillDir: string): BundledResources {
  const resources: BundledResources = {
    scripts: [],
    references: [],
    assets: [],
    other: [],
  };

  if (!fs.existsSync(skillDir)) return resources;

  const KNOWN_DIRS: Record<string, keyof Omit<BundledResources, 'other'>> = {
    scripts: 'scripts',
    references: 'references',
    assets: 'assets',
  };

  const entries = fs.readdirSync(skillDir, { withFileTypes: true });
  for (const entry of entries) {
    if (entry.name === 'SKILL.md') continue;

    if (entry.isDirectory()) {
      const category = KNOWN_DIRS[entry.name];
      if (category) {
        const subDir = path.join(skillDir, entry.name);
        const subEntries = fs.readdirSync(subDir);
        for (const sub of subEntries) {
          resources[category].push(`${entry.name}/${sub}`);
        }
      } else {
        const subDir = path.join(skillDir, entry.name);
        const subEntries = fs.readdirSync(subDir);
        for (const sub of subEntries) {
          resources.other.push(`${entry.name}/${sub}`);
        }
      }
    } else {
      resources.other.push(entry.name);
    }
  }

  return resources;
}

// ─── SkillLoader ──────────────────────────────────────────────────────────

/**
 * SkillLoader reads and validates skills from directories
 */
export class SkillLoader {
  private logger = getGlobalMatimoLogger();

  /**
   * Load all skills from a directory
   */
  loadSkillsFromDirectory(
    skillsDir: string,
    source: 'builtin' | 'user' | 'catalog' = 'user'
  ): SkillDefinition[] {
    const skills: SkillDefinition[] = [];

    if (!fs.existsSync(skillsDir)) {
      return skills;
    }

    const entries = fs.readdirSync(skillsDir, { withFileTypes: true });
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;

      const skillPath = path.join(skillsDir, entry.name, 'SKILL.md');
      if (!fs.existsSync(skillPath)) continue;

      try {
        this.logger.debug('SkillLoader: attempting to load skill', {
          name: entry.name,
          skillPath,
        });
        const skill = this.loadSkill(entry.name, skillsDir, source);
        if (skill) {
          skills.push(skill);
          this.logger.debug('SkillLoader: successfully loaded skill', {
            name: skill.name,
          });
        } else {
          this.logger.warn('SkillLoader: loadSkill returned null', {
            name: entry.name,
          });
        }
      } catch (err) {
        this.logger.error('SkillLoader: failed to load skill', {
          dir: entry.name,
          error: (err as Error).message,
          stack: (err as Error).stack,
          skillsDir,
        });
      }
    }

    return skills;
  }

  /**
   * Load a single skill by name
   */
  loadSkill(
    name: string,
    skillsDir: string,
    source: 'builtin' | 'user' | 'catalog' = 'user'
  ): SkillDefinition | null {
    // Validate name
    const nameValidation = validateSkillName(name);
    if (!nameValidation.valid) {
      throw new MatimoError(
        `Invalid skill name: ${nameValidation.error}`,
        ErrorCode.INVALID_SCHEMA
      );
    }

    const skillDir = path.join(skillsDir, name);
    const skillPath = path.join(skillDir, 'SKILL.md');

    if (!fs.existsSync(skillPath)) {
      throw new MatimoError(`Skill not found: ${skillPath}`, ErrorCode.TOOL_NOT_FOUND);
    }

    const content = fs.readFileSync(skillPath, 'utf-8');
    const parsed = parseSkillContent(content);

    if (parsed.error) {
      throw new MatimoError(`Failed to parse skill: ${parsed.error}`, ErrorCode.INVALID_SCHEMA);
    }

    const { frontmatter } = parsed;

    // Verify frontmatter name matches directory name
    if (frontmatter.name !== name) {
      throw new MatimoError(
        `Skill name "${frontmatter.name}" must match directory name "${name}"`,
        ErrorCode.INVALID_SCHEMA
      );
    }

    const resources = listBundledResources(skillDir);

    const skill: SkillDefinition = {
      name: frontmatter.name,
      description: frontmatter.description,
      version: frontmatter.version,
      license: frontmatter.license,
      compatibility: frontmatter.compatibility,
      allowedTools: frontmatter['allowed-tools'] as string[] | undefined,
      metadata: frontmatter.metadata,
      body: parsed.body,
      sections: parsed.sections,
      totalTokens: parsed.totalTokens,
      resources,
      source,
      _path: skillDir,
    };

    return skill;
  }

  /**
   * Load a skill resource file (scripts/, references/, assets/)
   */
  loadSkillResource(skillName: string, skillsDir: string, resourcePath: string): string {
    // Validate resource path (prevent traversal)
    if (/\.\.|\\/u.test(resourcePath) || /[\x00-\x1f]/.test(resourcePath)) {
      throw new MatimoError('Resource path contains invalid characters', ErrorCode.INVALID_SCHEMA);
    }

    const skillDir = path.join(skillsDir, skillName);
    const resourceFullPath = path.join(skillDir, resourcePath);

    // Verify path stays within skill directory
    const resolvedPath = path.resolve(resourceFullPath);
    const resolvedSkillDir = path.resolve(skillDir);
    if (
      !resolvedPath.startsWith(resolvedSkillDir + path.sep) &&
      resolvedPath !== resolvedSkillDir
    ) {
      throw new MatimoError('Resource path escapes the skill directory', ErrorCode.INVALID_SCHEMA);
    }

    if (!fs.existsSync(resourceFullPath)) {
      throw new MatimoError(`Resource file not found: ${resourcePath}`, ErrorCode.TOOL_NOT_FOUND);
    }

    try {
      return fs.readFileSync(resourceFullPath, 'utf-8');
    } catch (err) {
      throw new MatimoError(
        `Failed to read resource file: ${(err as Error).message}`,
        ErrorCode.EXECUTION_FAILED
      );
    }
  }
}

File: python/packages/core/src/matimo/core/skill_loader.py
"""
Skill Loader — loads and validates skills from the filesystem.

Mirrors: packages/core/src/core/skill-loader.ts

Implements agentskills.io specification with proper YAML parsing and
Pydantic validation.
"""
from __future__ import annotations

import logging
import re
from pathlib import Path

import yaml

from matimo.core.models import (
    BundledResources,
    ParsedSkill,
    SkillDefinition,
    SkillFrontmatter,
    SkillSummary,
)
from matimo.core.skill_content_parser import parse_skill_sections
from matimo.errors import ErrorCode, MatimoError

logger = logging.getLogger("matimo")

# ---------------------------------------------------------------------------
# Name validation
# ---------------------------------------------------------------------------

_VALID_NAME_RE = re.compile(r"^[a-z0-9]([a-z0-9-]*[a-z0-9])?$")
_CONSECUTIVE_HYPHENS_RE = re.compile(r"--")
_MAX_NAME_LENGTH = 64


def _validate_skill_name(name: str) -> str | None:
    """Return error string or None if valid."""
    if not name or not name.strip():
        return "Skill name is required"
    if len(name) > _MAX_NAME_LENGTH:
        return f"Skill name must be at most {_MAX_NAME_LENGTH} characters"
    if not _VALID_NAME_RE.match(name):
        return (
            "Skill name must contain only lowercase letters, numbers, and hyphens, "
            "and must not start or end with a hyphen"
        )
    if _CONSECUTIVE_HYPHENS_RE.search(name):
        return "Skill name must not contain consecutive hyphens"
    return None


# ---------------------------------------------------------------------------
# Frontmatter extraction
# ---------------------------------------------------------------------------


def _extract_frontmatter(
    content: str,
) -> tuple[SkillFrontmatter | None, str | None, str | None]:
    """
    Extract and validate YAML frontmatter from SKILL.md content.

    Returns ``(frontmatter, body, error)``.
    """
    if not content or not content.startswith("---"):
        return None, None, "Skill content must start with YAML frontmatter (---)"

    end_idx = content.index("---", 3) if "---" in content[3:] else -1
    if end_idx == -1:
        return None, None, "Skill content must have closing YAML frontmatter (---)"
    # Adjust because we searched from position 3
    end_idx = content.index("---", 3)

    fm_block = content[3:end_idx].strip()
    body = content[end_idx + 3 :].strip()

    try:
        parsed = yaml.safe_load(fm_block) or {}
    except yaml.YAMLError as exc:
        return None, None, f"Failed to parse YAML frontmatter: {exc}"

    if not isinstance(parsed, dict):
        return None, None, "Frontmatter must be a YAML mapping"

    # Normalise allowed-tools: space-delimited string → list
    at = parsed.get("allowed-tools")
    if isinstance(at, str):
        parsed["allowed-tools"] = at.split()

    # Validate required fields
    if not parsed.get("name"):
        return None, None, "Frontmatter validation failed: name is required"
    if not parsed.get("description"):
        return None, None, "Frontmatter validation failed: description is required"

    try:
        frontmatter = SkillFrontmatter.model_validate(parsed)
    except Exception as exc:
        return None, None, f"Frontmatter validation failed: {exc}"

    return frontmatter, body, None


# ---------------------------------------------------------------------------
# Public helpers
# ---------------------------------------------------------------------------


def parse_skill_content(content: str) -> ParsedSkill:
    """Parse YAML frontmatter from SKILL.md content."""
    frontmatter, body, error = _extract_frontmatter(content)

    if error:
        # Return a best-effort result
        fallback_body = ""
        idx = content.find("---", 3)
        if idx != -1:
            fallback_body = content[idx + 3 :].strip()
        return ParsedSkill(
            frontmatter=SkillFrontmatter(name="", description=""),
            body=fallback_body,
            raw=content,
        )

    assert frontmatter is not None
    assert body is not None

    parsed_content = parse_skill_sections(body)

    return ParsedSkill(
        frontmatter=frontmatter,
        body=body,
        raw=content,
        sections=parsed_content.sections,
        total_tokens=parsed_content.total_tokens,
    )


def extract_skill_metadata(
    content: str,
    source: str = "user",
) -> tuple[bool, SkillSummary | None, str | None]:
    """
    Extract ONLY metadata from SKILL.md without parsing body/sections.

    Optimised for listing skills — reads YAML frontmatter only.
    Returns ``(success, summary, error)``.
    """
    frontmatter, _, error = _extract_frontmatter(content)
    if error:
        return False, None, error

    assert frontmatter is not None
    return (
        True,
        SkillSummary(
            name=frontmatter.name,
            description=frontmatter.description,
            version=frontmatter.version,
            license=frontmatter.license,
            metadata=frontmatter.metadata,
            source=source,  # type: ignore[arg-type]
        ),
        None,
    )


# ---------------------------------------------------------------------------
# Bundled resources
# ---------------------------------------------------------------------------

_KNOWN_DIRS: dict[str, str] = {
    "scripts": "scripts",
    "references": "references",
    "assets": "assets",
}


def _list_bundled_resources(skill_dir: Path) -> BundledResources:
    resources = BundledResources()
    if not skill_dir.is_dir():
        return resources

    for entry in skill_dir.iterdir():
        if entry.name == "SKILL.md":
            continue

        if entry.is_dir():
            category = _KNOWN_DIRS.get(entry.name)
            sub_entries = [f.name for f in entry.iterdir() if not f.name.startswith(".")]
            items = [f"{entry.name}/{s}" for s in sub_entries]
            if category:
                getattr(resources, category).extend(items)
            else:
                resources.other.extend(items)
        else:
            resources.other.append(entry.name)

    return resources


# ---------------------------------------------------------------------------
# SkillLoader
# ---------------------------------------------------------------------------


class SkillLoader:
    """Reads and validates skills from directories."""

    def load_skills_from_directory(
        self,
        skills_dir: str,
        source: str = "user",
    ) -> list[SkillDefinition]:
        """Load all skills from a directory."""
        skills: list[SkillDefinition] = []
        skills_path = Path(skills_dir)

        if not skills_path.is_dir():
            return skills

        for entry in skills_path.iterdir():
            if not entry.is_dir():
                continue

            skill_path = entry / "SKILL.md"
            if not skill_path.is_file():
                continue

            try:
                skill = self.load_skill(entry.name, skills_dir, source)
                if skill:
                    skills.append(skill)
                    logger.debug("SkillLoader: loaded skill %s", skill.name)
            except Exception:
                logger.exception("SkillLoader: failed to load skill from %s", entry.name)

        return skills

    def load_skill(
        self,
        name: str,
        skills_dir: str,
        source: str = "user",
    ) -> SkillDefinition | None:
        """Load a single skill by name."""
        error = _validate_skill_name(name)
        if error:
            raise MatimoError(f"Invalid skill name: {error}", ErrorCode.INVALID_SCHEMA)

        skill_dir = Path(skills_dir) / name
        skill_path = skill_dir / "SKILL.md"

        if not skill_path.is_file():
            raise MatimoError(f"Skill not found: {skill_path}", ErrorCode.TOOL_NOT_FOUND)

        content = skill_path.read_text(encoding="utf-8")
        parsed = parse_skill_content(content)

        if not parsed.frontmatter.name:
            raise MatimoError(
                f"Failed to parse skill: invalid frontmatter in {skill_path}",
                ErrorCode.INVALID_SCHEMA,
            )

        if parsed.frontmatter.name != name:
            raise MatimoError(
                f'Skill name "{parsed.frontmatter.name}" must match directory name "{name}"',
                ErrorCode.INVALID_SCHEMA,
            )

        resources = _list_bundled_resources(skill_dir)

        allowed_tools = parsed.frontmatter.allowed_tools
        if isinstance(allowed_tools, str):
            allowed_tools = [allowed_tools]

        return SkillDefinition(
            name=parsed.frontmatter.name,
            description=parsed.frontmatter.description,
            version=parsed.frontmatter.version,
            license=parsed.frontmatter.license,
            compatibility=parsed.frontmatter.compatibility,
            allowed_tools=allowed_tools,
            metadata=parsed.frontmatter.metadata,
            body=parsed.body,
            sections=parsed.sections,
            total_tokens=parsed.total_tokens,
            resources=resources,
            source=source,  # type: ignore[arg-type]
            **{"_path": str(skill_dir)},
        )

    def load_skill_resource(
        self,
        skill_name: str,
        skills_dir: str,
        resource_path: str,
    ) -> str:
        """Load a skill resource file (scripts/, references/, assets/)."""
        # Validate resource path (prevent traversal)
        if ".." in resource_path or "\\" in resource_path:
            raise MatimoError(
                "Resource path contains invalid characters",
                ErrorCode.INVALID_SCHEMA,
            )
        # Reject control characters
        if any(0 <= ord(c) <= 0x1F for c in resource_path):
            raise MatimoError(
                "Resource path contains invalid characters",
                ErrorCode.INVALID_SCHEMA,
            )

        skill_dir = Path(skills_dir) / skill_name
        resource_full = (skill_dir / resource_path).resolve()

        # Verify path stays within skill directory
        if not str(resource_full).startswith(str(skill_dir.resolve())):
            raise MatimoError(
                "Resource path escapes the skill directory",
                ErrorCode.INVALID_SCHEMA,
            )

        if not resource_full.is_file():
            raise MatimoError(
                f"Resource file not found: {resource_path}",
                ErrorCode.TOOL_NOT_FOUND,
            )

        try:
            return resource_full.read_text(encoding="utf-8")
        except Exception as exc:
            raise MatimoError(
                f"Failed to read resource file: {exc}",
                ErrorCode.EXECUTION_FAILED,
            ) from exc
